text/0000-unsafe-aliased.md

```diff
@@ -0,0 +1,432 @@
+
+### Naming
+
+An earlier proposal suggested to call the type `UnsafeAliased`, since the type is not inherently tied to pinning.
+However, it is not possible to write sound wrappers around `UnsafeAliased` such that we can have aliasing `&mut MyType`. One has to use pinning for that: `Pin<&mut MyType>`.
+Because of that, the RFC proposes that we suggestively put pinning into the name of the type, so that people don't confuse it with a general mechanism for aliasing mutable references.
+It is more like the core primitive behind pinning, where whenever a type is pinned that is caused by an `UnsafePinned` field somewhere inside it.
+
+For instance, it may be tempting to use an `UnsafeAliased` type to mark a single field in some struct as "separately aliased", and then a `Mutex<Struct>` would acquire ownership of the entire struct except for that field.
+However, due to `mem::swap`, that would not be sound.
+One cannot hand out an `&mut` to such aliased memory as part of a safe-to-use abstraction -- except by using pinning.
+
+# Prior art
+[prior-art]: #prior-art
+
+This is somewhat like `UnsafeCell`, but for mutable instead of shared references.
+
+Adding something like this to Rust has been discussed many times throughout the years. Here are some links for recent discussions:
+- [Zulip, Nov 2022](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/UnsafeCellMut/near/308124014)
+- [IRLO, May 2023](https://internals.rust-lang.org/t/an-alternative-to-the-current-unpin-hack-unsafealias/18810)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- How do we transition code that relies on `Unpin` opting-out of aliasing guarantees to the new type? Here's a plan: define the `PhantomPinned` type as `UnsafePinned<()>`.
+  Places in the standard library that use `impl !Unpin for` and the generator lowering are adjusted to use `UnsafePinned` instead.
+  Then as long as nobody outside the standard library used the unstable `impl !Unpin for`, switching the `noalias`-opt-out to the `UnsafeUnpin` trait is actually backwards compatible with the (never explicitly supported) `Unpin` hack!
+- Relatedly, in which module should this type live?
+- Should this type `derive(Copy)`? `UnsafeCell` does not, which is unfortunate because it now means some people might use `T: Copy` as indication that there is no `UnsafeCell` in `T`.
+- `Unpin` [also affects the `dereferenceable` attribute](https://github.com/rust-lang/rust/pull/106180), so the same would happen for this type. Is that something we want to guarantee, or do we hope to get back `dereferenceable` when better semantics for it materialize on the LLVM side?
+- When should `UnsafePinned<T>` be `Send` or `Sync`?
+  Self-referential futures can be `Send` because once they are pinned, they cannot be moved, so certainly they cannot be sent -- in other words, `Send` only constrains the unpinned state of a type.
```

Comment:
> `Send` only constrains the unpinned state of a type

This sounds very incorrect to me. Of course, self-referential futures ***can*** be sent after being pinned. Sending is a logical property of moving *access* to a value between the thread, not a property of physical bit-copying of a value to a different location. (All threads share a common heap is a fundamental concept of shared-memory multi-threading, and Rust very much supports it.)

By sending a `Pin<Box<T>>` to another thread, such is done easily for pinned data, and *of course* requires `T: Send`. Also compiler checks around `async` blocks that determine which types are ever held onto across `.await` points and which arent does *of course **only*** address the soundness of `Send` implementations for self-referntial futures in their *pinned* state.

---

I don’t have a good answer to the (valid) unresolved question “When should `UnsafePinned<T>` be `Send` or `Sync`?” at this time, but I’d recommend removing this additional line of confused “explanation”.

```suggestion
- When should `UnsafePinned<T>` be `Send` or `Sync`?
```

Reply:
You are absolutely right, I have no idea what I was thinking when I wrote this.